### PR TITLE
Add static validation to middleware

### PR DESCRIPTION
**Pull request by katherinelc321**
### What this PR does
**Before this PR:**
Not enough static validation on the frontend
**After this PR:**
Added validation for subscription ID, resource name, resource group name
Jira: ARO-5736
Link to demo recording: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Deployment: The deployment process was considered and addressed if required
- [ ] Testing: New code requires new unit tests.
- [ ] Documentation: Is the documentation updated? Either in the doc located in focus area, in the README or in the code itself.
- [ ] Customers: Is this change affecting customers? Is the release plan considered?

### CODE REVIEW

File: frontend/middleware_validate.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+const (
+	wantedResourceProviderNamespace = "Microsoft.RedHatOpenShift"
+	resourceTypeHCPOpenshiftCluster = "HcpOpenShiftClusters"
+)
+
+var rxResourceGroupName = regexp.MustCompile(`^[-a-z0-9_().]{0,89}[-a-z0-9_()]$`)
+
+func MiddlewareValidate() MiddlewareFunc {
```
Comment (s-amann) [inline frontend/middleware_validate.go:26]: instead of wrapping the function creation, can we implement the `MiddlewareFunc` function signature? This aligns it with the other middleware we have

`func MiddlewareValidate(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {`

Comment (s-amann) [inline frontend/middleware_validate.go:26]: Since we have multiple middleware performing validation, can we rename this one to be `MiddlewareValidateStatic`?

File: frontend/middleware_validatestatic.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/Azure/ARO-HCP/internal/api"
+	"github.com/Azure/ARO-HCP/internal/api/arm"
+)
+
+const (
+	wantedResourceProviderNamespace = "Microsoft.RedHatOpenShift"
+	resourceTypeHCPOpenshiftCluster = "HcpOpenShiftClusters"
+)
+
+var rxResourceGroupName = regexp.MustCompile(`^[-a-z0-9_().]{0,89}[-a-z0-9_()]$`)
```
Comment (s-amann) [inline frontend/middleware_validatestatic.go:24]: where does this regex come from? Is there documentation we can comment regarding this?

File: frontend/middleware_validatestatic.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		valid := api.IsValid(subId)
+		if !valid {
+			arm.WriteError(w, http.StatusBadRequest, arm.CloudErrorCodeInvalidSubscriptionID, "", "The provided subscription identifier '%s' is malformed or invalid.", subId)
+			return
+		}
+	}
+
+	if resourceGroupName != "" {
+		if !rxResourceGroupName.MatchString(resourceGroupName) {
+			arm.WriteError(w, http.StatusBadRequest, arm.CloudErrorCodeResourceGroupNotFound, "", "Resource group '%s' could not be found.", resourceGroupName)
```
Comment (s-amann) [inline frontend/middleware_validatestatic.go:55]: this error message is misleading, we did not attempt a resourceGroup lookup to determine if it could be found. Can we rephrase this to align with the validation?

File: frontend/middleware_validatestatic.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/go-chi/chi/v5"
+	"github.com/go-logr/logr"
+
+	"github.com/Azure/ARO-HCP/internal/api"
+	"github.com/Azure/ARO-HCP/internal/api/arm"
+)
+
+const (
+	wantedResourceProviderNamespace = "Microsoft.RedHatOpenShift"
```
Comment (s-amann) [inline frontend/middleware_validatestatic.go:20]: This is already a constant in `internal/api/register.go` can we use it instead?

Comment (s-amann) [reply frontend/middleware_validatestatic.go:20]: The same applies for `resourceTypeHCPOpenshiftCluster`

File: frontend/middleware_validatestatic.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+	if resourceGroupName != "" {
+		if !rxResourceGroupName.MatchString(resourceGroupName) {
+			arm.WriteError(w, http.StatusBadRequest, arm.CloudErrorCodeResourceGroupNotFound, "", "Resource group '%s' could not be found.", resourceGroupName)
+			return
+		}
+	}
+
+	if resourceProviderNamespace != "" {
+		if resourceProviderNamespace != strings.ToLower(wantedResourceProviderNamespace) {
```
Comment (s-amann) [inline frontend/middleware_validatestatic.go:61]: nit: the check for empty string is redundant when we check if it equals a constant.

File: frontend/middleware_validatestatic.go (hunk 2 above)
Comment (katherinelc321) [reply frontend/middleware_validatestatic.go:24]: There are rules here: https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules#microsoftresources
They allow these but we don't need them since we only use the English alphabet
TitlecaseLetter,
ModifierLetter,
OtherLetter,

File: frontend/middleware_validatestatic.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+var rxResourceGroupName = regexp.MustCompile(`^[-a-z0-9_().]{0,89}[-a-z0-9_()]$`)
+
+func MiddlewareValidateStatic(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+
+	subId := chi.URLParam(r, "subscriptionId")
+	resourceGroupName := chi.URLParam(r, "resourceGroupName")
+	resourceProviderNamespace := chi.URLParam(r, "resourceProviderNamespace")
+	resourceType := chi.URLParam(r, "resourceType")
+	operationId := chi.URLParam(r, "operationId")
+	resourceName := chi.URLParam(r, "resourceName")
```
Comment (mbarnes) [inline frontend/middleware_validatestatic.go:29]: We're not using `github.com/go-chi/chi` for ARO-HCP since Go 1.22 has HTTP routing built in.
What you want here is [r.PathValue](https://pkg.go.dev/net/http@go1.22.3#Request.PathValue).
Also please use the `PathSegment*` constants in `frontend/const.go`.

File: frontend/middleware_validatestatic.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+	apiVersion := r.URL.Query().Get(APIVersionKey)
+
+	if r.URL.Path != strings.ToLower(r.URL.Path) {
+		if log, ok := r.Context().Value(contextKeyOriginalPath).(logr.Logger); ok {
```
Comment (mbarnes) [inline frontend/middleware_validatestatic.go:34]: @AldoFusterTurpin recently added type-safe functions for these context values.  Use `LoggerFromContext` here.

File: frontend/middleware_validatestatic.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+			log.Error(errors.New("LowerCase error"), "path was not lower case")
+		}
+		arm.WriteError(w, http.StatusInternalServerError, arm.CloudErrorCodeInternalServerError, "", "Internal server error.")
```
Comment (mbarnes) [inline frontend/middleware_validatestatic.go:37]: There's a helper function for this specific error since it's so frequest: `arm.WriteInternalServerError`

File: internal/api/utils.go [hunk 9]
```diff
@@ -33,3 +37,9 @@ func StringPtrSliceToStringSlice(s []*string) []string {
 	}
 	return out
 }
+
+// Check that a string conforms to the UUID format
+func IsValid(u string) bool {
+	_, err := gofrsuuid.FromString(u)
```
Comment (mbarnes) [inline internal/api/utils.go:43]: We're already importing another UUID module "`github.com/google/uuid`" so we should stick to one -- whichever is more capable.  I don't know much about "`github.com/gofrs/uuid`", but Google's module has a [Validate](https://pkg.go.dev/github.com/google/uuid#Validate) function that seems like it would work here.

**Review by mbarnes — CHANGES_REQUESTED** [review]
Looks good overall, just some minor requests to reduce code duplication and unnecessary imports.

File: frontend/middleware_validatestatic.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"errors"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/go-chi/chi/v5"
```
Comment (s-amann) [inline frontend/middleware_validatestatic.go:12]: we should avoid bringing `chi` in unless we have to. To get URLParams, we can use built in go `request.PathValue()` This may require us to break the patterns we currently have in `frontend.go` along with `const.go` to break out the "resourceProvider"

File: frontend/middleware_validatestatic.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+		}
+	}
+
+	if resourceProviderNamespace != strings.ToLower(api.ProviderNamespace) {
+		arm.WriteError(w, http.StatusBadRequest, arm.CloudErrorCodeInvalidResourceNamespace, "", "The resource namespace '%s' is invalid.", resourceProviderNamespace)
+		return
+	}
+
+	if resourceType != "" {
+		if resourceType != strings.ToLower(api.ResourceType) {
```
Comment (s-amann) [inline frontend/middleware_validatestatic.go:62]: We should probably remove this check altogether. ARM will not route unregistered resourceTypes to our endpoint. It will only route deployed resourceTypes in our ARM Manifest.

That being said, we can still validate if we'd like to, but all requests will not be for "HcpOpenShiftClusters". We could move this into its own validator, and have it be extensible to plug in different resourceTypes to validate, and only include in in the middleware chain it belongs in.

File: frontend/middleware_validatestatic.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+			arm.WriteError(w, http.StatusBadRequest, arm.CloudErrorCodeInvalidResourceType, "", "The resource type '%s' could not be found in the namespace '%s' for api version '%s'.", resourceType, resourceProviderNamespace, apiVersion)
+			return
+		}
+	}
+
+	if resourceName != "" {
+		if !rxResourceGroupName.MatchString(resourceName) {
+			arm.WriteError(w, http.StatusBadRequest, arm.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s/%s' under resource group '%s' was not found.", resourceProviderNamespace, resourceType, resourceName, resourceGroupName)
```
Comment (s-amann) [inline frontend/middleware_validatestatic.go:70]: This doesn't actually check if the resourceName exists within a resourceGroup. If we are trying to validate the resourceName, we should have an error message tailored to the check. Do the resourceGroup and resourceName have the same requirements?

File: frontend/middleware_validatestatic.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/go-logr/logr"
+
+	"github.com/Azure/ARO-HCP/internal/api"
+	"github.com/Azure/ARO-HCP/internal/api/arm"
+)
+
+// Referenced in https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules#microsoftresources
+var rxResourceGroupName = regexp.MustCompile(`^[-a-z0-9_().]{0,89}[-a-z0-9_()]$`)
+
+func MiddlewareValidateStatic(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
```
Comment (s-amann) [inline frontend/middleware_validatestatic.go:22]: please create some unit tests for the various conditionals used in this function.

**Review by s-amann — CHANGES_REQUESTED** [review]

File: frontend/middleware_validatestatic.go (hunk 10 above)
Comment (mbarnes) [reply frontend/middleware_validatestatic.go:12]: The `http.ServeMux` routes are set up with the resource provider namespace hard-coded into the path.  So if the namespace doesn't match, the request will 404 and not even reach this middleware.  For the resource type I agree with your comment below.

File: frontend/middleware_validatestatic.go (hunk 12 above)
Comment (katherinelc321) [reply frontend/middleware_validatestatic.go:70]: Each resource has their own requirements. The resource group one is a good upper limit.  I can also narrow it down to this from the api spec:
            "name": "hcpOpenShiftClusterName",
            "in": "path",
            "description": "Name of HCP cluster",
            "required": true,
            "type": "string",
            "minLength": 3,
            "maxLength": 24,
            "pattern": "^[a-zA-Z0-9-]{3,24}$"
          }

**Review by petrkotas — CHANGES_REQUESTED** [review]
Hi @katherinelc321 thank you for doing the extra validation. I agree it is important to do it right from the start. Overall the PR is on a good path with some minor tweaks needed. Would you please look at the comments ?

File: internal/api/utils.go [hunk 14]
```diff
@@ -33,3 +37,9 @@ func StringPtrSliceToStringSlice(s []*string) []string {
 	}
 	return out
 }
+
+// Check that a string conforms to the UUID format
+func IsValid(u string) bool {
+	err := uuid.Validate(u)
+	return err == nil
+}
```
Comment (mbarnes) [inline internal/api/utils.go:45]: Suggest either eliminating this function and just doing `if uuid.Validate(u) != nil` where needed, or else rename the function to `UUIDIsValid` to clarify _what_ is valid.

**Review by mbarnes — CHANGES_REQUESTED** [review]
Some more changes are needed, see comments below.

tl;dr The path segment constants you're adding are not being used in the route patterns, so those path values will not be set.  I make a case for why they may not be necessary anyway.

File: frontend/middleware_validatestatic.go [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+	if subId != "" {
+		if uuid.Validate(subId) != nil {
+			arm.WriteError(w, http.StatusBadRequest, arm.CloudErrorCodeInvalidSubscriptionID, "", "The provided subscription identifier '%s' is malformed or invalid.", subId)
+			return
+		}
+	}
+
+	if resourceGroupName != "" {
+		if !rxResourceGroupName.MatchString(resourceGroupName) {
+			arm.WriteError(w, http.StatusBadRequest, arm.CloudErrorCodeResourceGroupNotFound, "", "Resource group '%s' is invalid.", resourceGroupName)
```
Comment (s-amann) [inline frontend/middleware_validatestatic.go:35]: since we are not checking if the resource group exists, can error code be `arm.CloudErrorCodeInvalidResourceGroupName`? with the value being `"InvalidResourceGroupName"`?

**Review by mbarnes — APPROVED** [review]
LGTM now, thanks!

**Review by s-amann — APPROVED** [review]
Thank you for making those changes, lgtm!
